src/integer_mod_q/mat_zq/vector/dot_product.rs

```diff
@@ -0,0 +1,194 @@
+                self_entries.len(),
+                other_entries.len()
+            )));
+        }
+
+        // calculate dot product of vectors
+        let mut result = Z::ZERO;
+        for i in 0..self_entries.len() {
+            // sets result = result + self.entry[i] * other.entry[i] without cloned Z element
+            unsafe { fmpz_addmul(&mut result.value, &self_entries[i], &other_entries[i]) }
+        }
+
+        Ok(Zq::from_z_modulus(&result, &self.get_mod()))
+    }
+}
+
+#[cfg(test)]
+mod test_dot_product {
+
+    use super::{MatZq, Zq, Z};
+    use std::str::FromStr;
+
+    /// Check whether the dot product is calculated correctly for the combination:
+    /// `self`: row vector, `other`: row vector
+    #[test]
+    fn row_with_row() {
+        let vec_1 = MatZq::from_str("[[1,2,-3]] mod 5").unwrap();
+        let vec_2 = MatZq::from_str("[[1,3,2]] mod 5").unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, Zq::from_str("1 mod 5").unwrap());
+    }
+
+    /// Check whether the dot product is calculated correctly for the combination:
+    /// `self`: column vector, `other`: column vector
+    #[test]
+    fn column_with_column() {
+        let vec_1 = MatZq::from_str("[[1],[2],[-3]] mod 5").unwrap();
+        let vec_2 = MatZq::from_str("[[1],[3],[2]] mod 5").unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, Zq::from_str("1 mod 5").unwrap());
+    }
+
+    /// Check whether the dot product is calculated correctly for the combination:
+    /// `self`: row vector, `other`: column vector
+    #[test]
+    fn row_with_column() {
+        let vec_1 = MatZq::from_str("[[1,2,-3]] mod 5").unwrap();
+        let vec_2 = MatZq::from_str("[[1],[3],[2]] mod 5").unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, Zq::from_str("1 mod 5").unwrap());
+    }
+
+    /// Check whether the dot product is calculated correctly for the combination:
+    /// `self`: column vector, `other`: row vector
+    #[test]
+    fn column_with_row() {
+        let vec_1 = MatZq::from_str("[[1],[2],[-3]] mod 5").unwrap();
+        let vec_2 = MatZq::from_str("[[1,3,2]] mod 5").unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, Zq::from_str("1 mod 5").unwrap());
+    }
+
+    /// Check whether the dot product is calculated correctly with large numbers
+    #[test]
+    fn large_numbers() {
+        let vec_1 = MatZq::from_str(&format!("[[1,-1,{}]] mod {}", i64::MAX, u64::MAX)).unwrap();
+        let vec_2 = MatZq::from_str(&format!("[[1,{},1]] mod {}", i64::MIN, u64::MAX)).unwrap();
+        let cmp = Z::from(-1) * Z::from(i64::MIN) + Z::from(i64::MAX) + Z::ONE;
+        let cmp = Zq::try_from_z_z(&cmp, &Z::from(u64::MAX)).unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, cmp);
+    }
```

Comment:
I would suggest to choose the modulus larger so that the result is also large.
Maybe also just one vector with large numbers.
Currently, the result of the dot product is 1 after applying the modulus.

You can also replace `Z::from(-1)` with just `-1` and `Z::ONE` with `1`